core/vmx_ops.asm

```diff
@@ -0,0 +1,358 @@
+
+function asm_invept, 2
+    invept reg_arg1, [reg_arg2]
+    vmx_check
+    ret
+
+function asm_vmxon, 1
+    vmxon [reg_arg1]
+    vmx_check
+    ret
+
+function asm_vmxoff, 0
+    vmxoff
+    vmx_check
+    ret
+
+function asm_vmclear, 1
+    vmclear [reg_arg1]
+    vmx_check
+    ret
+
+function asm_vmptrld, 1
+    vmptrld [reg_arg1]
+    vmx_check
+    ret
+
+function asm_vmptrst, 1
+    vmptrst [reg_arg1]
+    vmx_check
+    ret
+
+function asm_vmread, 1
+    xor reg_ret, reg_ret
+    vmread reg_ret, reg_arg1
+    ret
+
+function asm_vmwrite, 2
+    vmwrite reg_arg1, reg_arg2
+    ret
+
+function asm_vmxrun, 2
+%ifidn __BITS__, 64
+    pushfq
+    push r8
+    push r9
+    push r10
+    push r11
+    push r12
+    push r13
+    push r14
+    push r15
+    push rcx
+    push rdx
+    push rsi
+    push rdi
+    push rbp
+    push rax
+    push rbx
+    ; write host rsp
+    mov ebx, 6C14h
+    mov rax, rsp
+    sub rax, 8h
+    vmwrite rbx, rax
+    pop rbx
+    pop rax
+    push rax
+    push rbx
+    ; push the state
+    push reg_arg1
+    cmp reg_arg2_16, 1h
+    mov rax, rcx
+    mov rcx, [rax + vcpu_state._rcx]
+    mov rdx, [rax + vcpu_state._rdx]
+    mov rbx, [rax + vcpu_state._rbx]
+    mov rbp, [rax + vcpu_state._rbp]
+    mov rsi, [rax + vcpu_state._rsi]
+    mov rdi, [rax + vcpu_state._rdi]
+    mov r8,  [rax + vcpu_state._r8]
+    mov r9,  [rax + vcpu_state._r9]
+    mov r10, [rax + vcpu_state._r10]
+    mov r11, [rax + vcpu_state._r11]
+    mov r12, [rax + vcpu_state._r12]
+    mov r13, [rax + vcpu_state._r13]
+    mov r14, [rax + vcpu_state._r14]
+    mov r15, [rax + vcpu_state._r15]
+    mov rax, [rax + vcpu_state._rax]
+    je .resume
+    vmlaunch
+    jmp .exit_entry_fail
+.resume:
+    vmresume
+    jmp .exit_entry_fail
+.exit_entry:
+    push rdi
+    mov rdi, [rsp+8]
+    mov [rdi + vcpu_state._rax], rax
+    mov [rdi + vcpu_state._rcx], rcx
+    mov [rdi + vcpu_state._rdx], rdx
+    pop rcx
+    mov [rdi + vcpu_state._rbx], rbx
+    mov [rdi + vcpu_state._rbp], rbp
+    mov [rdi + vcpu_state._rsi], rsi
+    mov [rdi + vcpu_state._rdi], rcx
+    mov [rdi + vcpu_state._r8], r8
+    mov [rdi + vcpu_state._r9], r9
+    mov [rdi + vcpu_state._r10], r10
+    mov [rdi + vcpu_state._r11], r11
+    mov [rdi + vcpu_state._r12], r12
+    mov [rdi + vcpu_state._r13], r13
+    mov [rdi + vcpu_state._r14], r14
+    mov [rdi + vcpu_state._r15], r15
+.exit_entry_fail:
+    ; pop the state
+    pop rbx
+    pop rbx
+    pop rax
+    pop rbp
+    pop rdi
+    pop rsi
+    pop rdx
+    pop rcx
+    pop r15
+    pop r14
+    pop r13
+    pop r12
+    pop r11
+    pop r10
+    pop r9
+    pop r8
+    vmx_check
+    popfq
+    ret
+%else
+    pushfd
+    push ecx
+    push edx
+    push esi
+    push edi
+    push ebp
+    push eax
+    push ebx
+    ; write host rsp
+    mov ebx, 6C14h
+    mov eax, esp
+    sub eax, 4h
+    vmwrite ebx, eax
+    pop ebx
+    pop eax
+    push eax
+    push ebx
+    ; push the state
+    push reg_arg1
```

Comment:
Here we need to copy the `state` pointer to `eax` as the original 32-bit MASM code used to do:

```
mov eax, reg_arg1
push eax
```

Otherwise, since `reg_arg1 == ecx != eax`, we'll be loading ECX with the wrong data at line 317. And we can't fix that by simply replacing `eax` with `reg_arg1` in lines 317-323, because line 317 will modify ECX and thus `reg_arg1`.